### PR TITLE
Terraform environments

### PR DESCRIPTION
**Pull request by AdamProbert**
Let's rewrite the terraform code, that's a good idea! 🤦 

This pr aims to restructure the entirety of our terraform code. Previously this was built in a very quick, rudimentary fashion. It had the following problems:
- Only one machine could run `terraform apply` (due to state stored locally)
- None of our resources were modularised leading to much duplication
- There was only one available environment leading to devs overwriting each others work
- We couldn't approve deployments to main

This PR aims to address all of these issues:
- Utilise `terraform workspace` to distinguish between environments
- Integrates closely with git branches to deploy duplicate infrastructure for each github branch
- Stores the terraform state and backend in AWS S3 with a complimenting lockfile in DynamoDB
- Created two seperate workflows, one for production (main) and one for everything else. The production workflow has an added step for manually approving terraform deployments. This should stop and accidental infrastructure changes.
- Modularised 70% of our resources for deduplication and to ease extensibility.
- Updates to the README to serve as documentation

### CODE REVIEW

File: terraform/pipeline/glue.tf [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+module "estimate_2021_jobs_job" {
+  source          = "../modules/glue-job"
+  script_name     = "estimate_2021_jobs.py"
+  glue_role       = aws_iam_role.sfc_glue_service_iam_role
+  resource_bucket = module.pipeline_resources
+  datasets_bucket = module.datasets_bucket
+
+  job_parameters = {
```
Comment (garypolzin) [inline terraform/pipeline/glue.tf:71]: Not sure if the parameters are right on this one?

**Review by garypolzin — APPROVED** [review]
impressive work!
can't see anything which stands out as being wrong or unusual

File: terraform/pipeline/glue.tf (hunk 1 above)
Comment (AdamProbert) [reply terraform/pipeline/glue.tf:71]: Oh good spot, it should just be:
```
    "--prepared_locations_source" = ""
    "--destination"               = ""
````
right?

Comment (AdamProbert) [reply terraform/pipeline/glue.tf:71]: Have updated 👍

File: terraform/modules/glue-crawler/glue-crawler.tf [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    delete_behavior = "DEPRECATE_IN_DATABASE"
+    update_behavior = "UPDATE_IN_DATABASE"
+  }
+
+  configuration = jsonencode(
+    {
+      "Version" : 1.0,
+      "Grouping" = {
+        "TableLevelConfiguration" = 3,
+        "TableGroupingPolicy" : "CombineCompatibleSchemas"
```
Comment (LauraZM9) [inline terraform/modules/glue-crawler/glue-crawler.tf:25]: Is there a reason for which you use both `=` and `:`? If not, I feel this should be more consistent.

**Review by LauraZM9 — CHANGES_REQUESTED** [review]

File: terraform/pipeline/step-function.tf [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+      "Next": "Success"
+    },
+    "Success": {
+      "Type": "Succeed"
+    },
+    "Fail": {
+      "Type": "Fail"
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (AdamProbert) [reply terraform/pipeline/step-function.tf]: Ahh nice, I did have a brief look for something like this. Have updated 🚀

File: README.md [hunk 4]
```diff
@@ -93,14 +93,29 @@ For verbose output add `-v` to the end of the command.
...
+When creating a new git branch and pushing to the remote repository a CircleCi workflow will automatically trigger. <br>
+One of the steps in this workflow is to deploy terraform. You can find the full CircleCi configuration inside [.circleci/config.yml](.circleci/config.yml).
+Once the workflow has completed AWS will contain all the infrastructure required to run the pipeline and all associated glue jobs.
+
+> ❗ **When merging with the main branch**: The workflow will run here too. There is a mandatory, manual approval step required here. Please read the output of `terraform plan`. Ensure this is correct, then give your approval. The workflow will complete and the main (production) infrastructure will be updated. The main branch workflows can be found [here](https://app.circleci.com/pipelines/github/NMDSdevopsServiceAdm/DataEngineering?branch=main&filter=all).
+
```
[the remark this answers is not in the document]
Comment (AdamProbert) [reply README.md:108]: Yep, good shout. Have updated!

File: README.md [hunk 5]
```diff
@@ -126,6 +144,28 @@ terraform plan
 ```
 terraform apply
 ```
+<br>
+
+### Destroying Terraform
+
+“What goes up must come down.” - Isaac Newton
```
Comment (LauraZM9) [inline README.md:151]: Inspiring!

File: README.md [hunk 6]
```diff
@@ -126,6 +144,28 @@ terraform plan
...
+
+Nobody wants a bunch of infrastructure left lying around, unused, it adds cognitive load, confusion and possible expense.
+
+To remove Terraform generated infrastructure first ensure you are in the correct directory working on the correct workspace.
+
+```
+cd terraform/pipeleine
```
Comment (LauraZM9) [inline README.md:158]: you mean "pipeline"

**Review by LauraZM9 — APPROVED** [review]
Great job, Adam! :tada:

File: README.md (hunk 6 above)
Comment (AdamProbert) [reply README.md:158]: ah of course I did, thank you!

**Pull request merged**